### PR TITLE
DRILL-5068: Create sys.profiles and sys.profiles_json tables

### PR DESCRIPTION
**Pull request by kkhatua**
(Thanks to Hongze Zhang for the initial work)
Introduced a non-distributed (i.e. the foreman Drillbit executes a single fragment query) system table to list the completed profiles that are visible to it in the persistent store. This allows for querying the profiles on the following common metrics:
    queryId
    startTime
    foreman
    fragments
    user
    queue
    planTime
    queueTime
    executeTime
    totalTime
    state
    query

An additional system table - `sys.profiles_json` has also been introduced. This only has the following fields:
    queryId
    json
This allows for more detail metrics to be queried on the profile as a JSON document. Users are expected to use `sys.profiles` to identify profiles on the available metrics (such as long executeTime or fragments), and use the queryId to find the corresponding JSON document representation for further drill down into the profiles.

Here is what a sample output looks like:
```
[root@kk127 ~]# /opt/drill/apache-drill-1.13.0/bin/sqlline -u "jdbc:drill:drillbit=`hostname -i`"
...
apache drill 1.13.0-SNAPSHOT 
"just drill it"
0: jdbc:drill:drillbit=10.10.100.127> select * from sys.profiles where startTime < date '2017-12-18' limit 10;
+---------------------------------------+--------------------------+---------------+------------+------------+--------+-----------+------------+--------------+------------+------------+--------------------------------------------------------+
|                queryId                |        startTime         |    foreman    | fragments  |    user    | queue  | planTime  | queueTime  | executeTime  | totalTime  |   state    |                         query                          |
+---------------------------------------+--------------------------+---------------+------------+------------+--------+-----------+------------+--------------+------------+------------+--------------------------------------------------------+
| 25cdcede-410d-6399-97f9-00ec2048f247  | 2017-12-14 08:06:57.062  | kk127.qa.lab  | 1          | anonymous  | -      | 35        | 0          | 7718         | 7753       | COMPLETED  | select queryId from profiles limit 1                   |
| 25cdd1a2-7928-bcf9-a8cd-8e0fa653e2ce  | 2017-12-14 07:55:09.551  | kk127.qa.lab  | 1          | anonymous  | -      | 41        | 0          | 7580         | 7621       | COMPLETED  | select queryId from profiles limit 1                   |
| 25cdd1b6-3a56-e2b3-a6a1-4d1006ec8fac  | 2017-12-14 07:54:49.403  | kk127.qa.lab  | 1          | anonymous  | -      | 77        | 0          | 7452         | 7529       | COMPLETED  | select queryId from profiles limit 1                   |
| 25cdd1bd-963f-7a57-6a2d-cf7a3eac3f00  | 2017-12-14 07:54:41.065  | kk127.qa.lab  | 0          | anonymous  | -      | 0         | 0          | 0            | 0          | FAILED     | select profileJson from profiles limit 1               |
| 25cdd1e5-de23-0cfe-5c01-86621b31d8d5  | 2017-12-14 07:54:01.576  | kk127.qa.lab  | 1          | anonymous  | -      | 1184      | 0          | 10860        | 12044      | COMPLETED  | select profileJson from profiles_json limit 1          |
| 25cdd1e7-7c3a-5a25-0860-a86246bbc9ed  | 2017-12-14 07:54:00.186  | kk127.qa.lab  | 1          | anonymous  | -      | 102       | 0          | 231          | 333        | COMPLETED  | use sys                                                |
| 25cdd1f0-4473-55aa-a5eb-c85194321118  | 2017-12-14 07:53:51.966  | kk127.qa.lab  | 0          | anonymous  | -      | 0         | 0          | 0            | 0          | FAILED     | select profileJson from profiles_json limit 1          |
| 25cdd442-04cc-65fe-7a25-ac075fa22467  | 2017-12-14 07:43:57.251  | kk127.qa.lab  | 1          | anonymous  | -      | 40        | 0          | 10553        | 10593      | COMPLETED  | select profileJson from profiles_json limit 1          |
| 25cdd7c8-614d-8bcc-3b89-08df12761a3a  | 2017-12-14 07:28:55.359  | kk127.qa.lab  | 1          | anonymous  | -      | 170       | 0          | 10508        | 10678      | COMPLETED  | select count(*) from (select * from profiles_json)     |
| 25cdd7e0-34ea-9040-ab2a-ef8c6a136440  | 2017-12-14 07:28:31.464  | kk127.qa.lab  | 0          | anonymous  | -      | 0         | 0          | 0            | 0          | FAILED     | select count(*) from (pwdselect * from profiles_json)  |
+---------------------------------------+--------------------------+---------------+------------+------------+--------+-----------+------------+--------------+------------+------------+--------------------------------------------------------+
10 rows selected (8.336 seconds)
0: jdbc:drill:drillbit=10.10.100.127> select * from sys.profiles_json where queryId = '25cdd1e7-7c3a-5a25-0860-a86246bbc9ed';
+---------+------+
| queryId | json |
+---------+------+
| 25cdd1e7-7c3a-5a25-0860-a86246bbc9ed | {"id":{"part1":2724064141780867621,"part2":603667490114619885},"type":1,"start":1513238040186,"end":1513238040519,"query":"use sys","foreman":{"address":"kk127.qa.lab","userPort":31010,"controlPort":31011,"dataPort":31012,"version":"1.12.0-SNAPSHOT","state":0},"state":2,"totalFragments":1,"finishedFragments":0,"fragmentProfile":[{"majorFragmentId":0,"minorFragmentProfile":[{"state":3,"minorFragmentId":0,"operatorProfile":[{"inputProfile":[{"records":1,"batches":2,"schemas":1}],"operatorId":0,"operatorType":26,"setupNanos":0,"processNanos":67070036,"peakLocalMemoryAllocated":4530688,"waitNanos":0},{"inputProfile":[{"records":0,"batches":0,"schemas":0}],"operatorId":0,"operatorType":13,"setupNanos":0,"processNanos":0,"peakLocalMemoryAllocated":4530176,"waitNanos":0},{"inputProfile":[{"records":1,"batches":1,"schemas":1}],"operatorId":0,"operatorType":13,"setupNanos":0,"processNanos":15787437,"peakLocalMemoryAllocated":4530176,"metric":[{"metricId":0,"longValue":41}],"waitNanos":18796208}],"startTime":1513238040329,"endTime":1513238040482,"memoryUsed":0,"maxMemoryUsed":9060864,"endpoint":{"address":"kk127.qa.lab","userPort":31010,"controlPort":31011,"dataPort":31012,"version":"1.12.0-SNAPSHOT","state":0},"lastUpdate":1513238040487,"lastProgress":1513238040487}]}],"user":"anonymous","optionsJson":"[ ]","planEnd":1513238040288,"queueWaitEnd":1513238040288,"totalCost":0.0,"queueName":"-"} |
+---------+------+
1 row selected (11.667 seconds)
```

### CODE REVIEW

**Comment by kkhatua** [comment]
@sohami 
Applied the checks for security to deny unauthorized access.
This is a screenshot with an admin user (`mapr`) querying the table.
adminuser

This is a screenshot with a non-admin user (`kuser1`) querying the table.
nonadminuser

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+          ExecConstants.ADMIN_USERS_VALIDATOR.getAdminUsers(options),
+          ExecConstants.ADMIN_USER_GROUPS_VALIDATOR.getAdminUserGroups(options))) {
+      return false;
+    }
+
+    //Passed checks
+    return true;
+  }
+
+  private Iterator<ProfileInfo> iterateProfileInfo(FragmentContext context) {
```
Comment (sohami) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: input parameter `context` is not used anywhere.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    try {
+      PersistentStore<UserBitShared.QueryProfile> profiles = profileStoreContext.getCompletedProfileStore();
+
+      return transform(getAuthorizedProfiles(profiles.getAll(), queryingUsername, isAdmin));
+
+    } catch (Exception e) {
+      logger.error(String.format("Unable to get persistence store: %s, ", profileStoreContext.getProfileStoreConfig().getName(), e));
```
Comment (sohami) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: `Exception e` should be an argument to `logger.error` not `String.format`

File: exec/java-exec/src/main/java/org/apache/drill/exec/ops/FragmentContext.java [hunk 3]
```diff
@@ -230,6 +230,10 @@ public DrillbitContext getDrillbitContext() {
     return context;
   }
 
+  public QueryContext getQueryContext() {
```
[the remark this answers is not in the document]
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/ops/FragmentContext.java]: I was using it to retrieve the user name via `queryContext.getQueryUserName()` but I can use the fragmentContext too. +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+import java.util.Map.Entry;
+
+/**
+ * DRILL-5068: Add a new system table for completed profiles
+ */
+public class ProfileIterator implements Iterator<Object> {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ProfileIterator.class);
+
+  private final QueryProfileStoreContext profileStoreContext;
+  private final Iterator<?> itr;
```
[the remark this answers is not in the document]
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: Legacy. I was originally trying to have the same iterator function for both JSON and tabular listing. +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        .getProfileStoreContext();
+
+    queryingUsername = context.getQueryUserName();
+    isAdmin = hasAdminPrivileges(context.getQueryContext());
+
+    itr = iterateProfileInfo(context);
+  }
+
+  private boolean hasAdminPrivileges(QueryContext queryContext) {
+    if (queryContext == null) {
```
[the remark this answers is not in the document]
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: Shouldn't be required any more.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java (hunk 1 above)
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: Fixed. +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java (hunk 2 above)
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: I was trying to format the error message to indicate the issue., but I realize I might be hiding the actual error message. +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        authorizedProfiles.add(profileKVPair);
+      }
+    }
+    return authorizedProfiles.iterator();
+  }
+
+  /**
+   * Iterating persistentStore as a iterator of {@link org.apache.drill.exec.store.sys.ProfileIterator.ProfileInfo}.
+   */
+  private Iterator<ProfileInfo> transform(Iterator<Map.Entry<String, UserBitShared.QueryProfile>> all) {
```
[the remark this answers is not in the document]
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileJsonIterator.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * DRILL-5068: Add a new system table for completed profiles (JSON)
+ */
+public class ProfileJsonIterator implements Iterator<Object> {
```
[the remark this answers is not in the document]
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileJsonIterator.java]: Reworked the code and tested. +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    return itr.next();
+  }
+
+  @Override
+  public void remove() {
+    throw new UnsupportedOperationException();
+  }
+
+  public static class ProfileInfo {
+    private static final String UnknownValue = "UNKNOWN";
```
[the remark this answers is not in the document]
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+    } catch (Exception e) {
+      logger.error(String.format("Unable to get persistence store: %s, ", profileStoreContext.getProfileStoreConfig().getName(), e));
+      return Iterators.singletonIterator(ProfileInfo.getDefault());
+    }
+  }
+
+  //Returns an iterator for authorized profiles
+  private Iterator<Entry<String, QueryProfile>> getAuthorizedProfiles (
+      Iterator<Entry<String, QueryProfile>> allProfiles, String username, boolean isAdministrator) {
```
[the remark this answers is not in the document]
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/sys/ProfileIterator.java]: +1 . Moved  the authorizedProfiles method to base class.

**Comment by kkhatua** [comment]
@sohami made the changes. You can review this.

**Comment by sohami** [comment]
Thanks for the changes. +1 LGTM.

**Pull request closed without merging**